### PR TITLE
[RLlib] Fixed and improved stateful model training on offline data.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Description
Using stateful models in Offline RL is an important feature and the major prerequisites for this feature are already implemented in RLlib's stack. However, some minor changes are needed to indeed train such models in the new stack. This PR implements these minor but important changes at different locations in the code:

1. It introduces the `STATE_OUT` key in the outputs of `BC`'s `forward` function to make the next hidden state available to the connectors and loss function. 
2. It adds in the `AddStatesFromEpisodesToBatch` the initial state to the batch for offline data.
3. It adds in `MARWIL` a burn-in for the state that can be controlled via `burnin_len`.
4. It generates sequence sampling in the `OfflinePreLearner` dependent on the `max_seq_len`, `lookback_len` and `burnin_len`.
5. It fixes multiple smaller bugs in `OfflineEnvRunner` when recording from class-referenced environments, in `offline_rl_with_image_data.py` and `cartpole_recording.py` examples when loading the `RLModule` from checkpoint.
6. It fixes the use of `explore=True` in evaluation of Offline RL tests and examples.
7. It adds recorded expert data from `StatelessCartPole` to the `s3://ray-example-data/rllib/offline-data/statelesscartpole`
8. It adds a test for learning on a single episode and a single batch from recorded stateful expert data. Adds also a test to use instead of recorded states the initial states for sequences.
9. Adds a new config parameter `prelearner_use_recorded_module_states` to either use recorded states from the data (`True`) or use the initial state from the `RLModule` (`False`).

## Additional information
The only API change is the introduction of a `burnin_len` to the `MARWIL/BC` config.

>__Note:__ Stateful model training is only possible in `BC` and `MARWIL` so far for Offline RL. For `IQL` and `CQL` these changes have to be initiated through the off-policy algorithms (`DQN/SAC`) and for these all the buffers need to provide sequence sampling which is implemented right now solely in the `EpisodeReplayBuffer`. Therefore a couple of follow-up PRs need to be produced:

1. Introduce sequence sampling to `PrioritizedEpisodeReplayBuffer`.
2. Introduce sequence sampling to `MultiAgentEpisodeReplayBuffer`.
3. Introduce sequence sampling to `MultiAgentPrioritizedEpisodeReplayBuffer`.
4. Introduce stateful model training in `SAC`.
5. Introduce stateful model training to `IQL/CQL`.

### CODE REVIEW

File: rllib/examples/offline_rl/cartpole_recording.py [hunk 1]
```diff
@@ -120,9 +125,9 @@
 
     # Configure the algorithm for offline recording.
     config.offline_data(
-        output="local:///tmp/cartpole/",
+        output="local:///home/simon/git-projects/ray/cartpole/",
```
[the remark this answers is not in the document]
Comment (simonsays1980) [reply rllib/examples/offline_rl/cartpole_recording.py]: Yeah this is still WIP. These will change.

File: rllib/offline/offline_env_runner.py [hunk 2]
```diff
@@ -57,8 +57,17 @@ def __init__(self, *, config: AlgorithmConfig, **kwargs):
...
+        if self.env:
+            # Set the subdir (environment specific).
+            if isinstance(self.env, str):
+                # `env` is a string.
+                self.subdir_path = self.env.lower()
+            else:
+                # `env`` is a class or callable we use its class name.
+                self.subdir_path = self.env.unwrapped.envs[
+                    0
+                ].unwrapped.__class__.__name__.lower()
```
[the remark this answers is not in the document]
Comment (simonsays1980) [reply rllib/offline/offline_env_runner.py]: Great catch. But, this case should be handled in the super class because if `SingleAgentEnvRunner._make_env()` did not succeed it should raise an error there.

File: rllib/algorithms/bc/torch/default_bc_torch_rl_module.py [hunk 3]
```diff
@@ -37,9 +37,22 @@ def setup(self):
...
         """Generic BC forward pass (for all phases of training/evaluation)."""
+        outputs = {}
         # Encoder embeddings.
         encoder_outs = self._encoder(batch)
         # Action dist inputs.
-        return {
-            Columns.ACTION_DIST_INPUTS: self._pi_head(encoder_outs[ENCODER_OUT]),
-        }
+        outputs.update(
+            {Columns.ACTION_DIST_INPUTS: self._pi_head(encoder_outs[ENCODER_OUT])}
```
Comment (pseudo-rnd-thoughts) [inline rllib/algorithms/bc/torch/default_bc_torch_rl_module.py]: `outputs = {Columns.ACTION_DIST_INPUTS: self._pi_head(encoder_outs[ENCODER_OUT])}` rather than updating

File: rllib/examples/offline_rl/cartpole_recording.py [hunk 4]
```diff
@@ -147,13 +155,25 @@
...
+        # Note, to load only a component of a checkpoint, specify the
+        # `component` argument and provide the path to the component.
+        component=(
+            f"{COMPONENT_LEARNER_GROUP}"
+            f"/{COMPONENT_LEARNER}"
+            f"/{COMPONENT_RL_MODULE}"
+        ),
     )
 
     # Evaluate over 10 iterations and record the data.
```
Comment (pseudo-rnd-thoughts) [inline rllib/examples/offline_rl/cartpole_recording.py:175]: Update to 50

File: rllib/offline/tests/test_offline_rl_stateful.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+
+        # Read the episodes and decode them.
+        episodes = [
+            SingleAgentEpisode.from_state(
+                msgpack.unpackb(state, object_hook=mnp.decode)
+            )
+            for state in batch["item"]
+        ][:1]
+        # Assert the episode has a decent return.
+        assert episodes[0].get_return() > 350.0, "Return must be >350.0"
```
Comment (pseudo-rnd-thoughts) [inline rllib/offline/tests/test_offline_rl_stateful.py:114]: Have you checked that every episodes in the training data that this is true?

File: rllib/offline/tests/test_offline_rl_stateful.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+        # Evaluate the updated policy for 5 episodes.
+        eval_episodes = self.algo.eval_env_runner_group.foreach_env_runner(
+            self._remote_eval_episode_fn,
+            local_env_runner=False,
+        )
+        # Assert the eval return is decent.
+        self.assertGreaterEqual(
+            np.mean([ep.get_return() for ep in eval_episodes[0]]),
+            100.0,
+            "Eval return must be >100.0",
```
Comment (pseudo-rnd-thoughts) [inline rllib/offline/tests/test_offline_rl_stateful.py]: Could you include the actual returns in this error message to help future debugging / reviewing log files

File: rllib/offline/tests/test_offline_rl_stateful.py [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+
+        eval_episodes = self.algo.eval_env_runner_group.foreach_env_runner(
+            self._remote_eval_episode_fn,
+            local_env_runner=False,
+        )
+        # Assert the eval return is decent.
+        self.assertGreaterEqual(
+            np.mean([ep.get_return() for ep in eval_episodes[0]]),
+            100.0,
+            "Eval return must be >100.0",
```
Comment (pseudo-rnd-thoughts) [inline rllib/offline/tests/test_offline_rl_stateful.py]: Add eval-episode return data for debugging

File: rllib/offline/offline_env_runner.py (hunk 2 above)
Comment (pseudo-rnd-thoughts) [inline rllib/offline/offline_env_runner.py]: Are you sure we want to lower this string? Isn't it more helpful to have the full correct class name? Repeat for above

[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply rllib/offline/offline_env_runner.py]: Can we raise an error nonetheless in case there is a change upstream

File: rllib/offline/offline_prelearner.py [hunk 8]
```diff
@@ -326,6 +300,70 @@ def _validate_episodes(
                 cleaned_episodes.add(eps)
         return cleaned_episodes
 
+    def _remove_states_from_episodes(
+        self,
+        episodes: List[SingleAgentEpisode],
```
Comment (pseudo-rnd-thoughts) [inline rllib/offline/offline_prelearner.py:305]: Could this support multi-agent episodes easily?

File: rllib/offline/offline_prelearner.py [hunk 9]
```diff
@@ -326,6 +300,70 @@ def _validate_episodes(
...
+            and not self.config.prelearner_use_recorded_module_states
+        ):
+            episodes = self._remove_states_from_episodes(episodes)
+
+        # Add the episodes to the buffer.
+        self.episode_buffer.add(episodes)
+
+        # Sample from the buffer.
+        batch_length_T = (
+            self.config.model_config.get("max_seq_len", 0)
```
Comment (pseudo-rnd-thoughts) [inline rllib/offline/offline_prelearner.py:350]: 1. Should the default be `None` rather than `0`? 
2. Does this work when the module is stateful but max_seq_len is 0 or None?

File: rllib/offline/offline_prelearner.py [hunk 10]
```diff
@@ -326,6 +300,70 @@ def _validate_episodes(
...
+            if self._module.is_stateful()
+            else None
+        )
+
+        return self.episode_buffer.sample(
+            num_items=self.config.train_batch_size_per_learner,
+            batch_length_T=batch_length_T,
+            n_step=self.config.get("n_step", 1) or 1,
```
Comment (pseudo-rnd-thoughts) [inline rllib/offline/offline_prelearner.py]: Should this `or 1` exist as `self.config.get` will throw an error if `None` rather than go to the `or 1`?

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
Overall looks good, just some high-level questions about several components

File: rllib/offline/offline_prelearner.py (hunk 8 above)
Comment (simonsays1980) [reply rllib/offline/offline_prelearner.py:305]: Great question: We do not support multi-agent offline RL (yet). When we do, we have to rewrite, because we need to iterate over single-agent episodes inside of multi-agent episodes

File: rllib/offline/offline_prelearner.py (hunk 10 above)
Comment (simonsays1980) [reply rllib/offline/offline_prelearner.py]: Great catch! The `or 1` is obsolete here. Must be a relict.

File: rllib/offline/offline_env_runner.py (hunk 2 above)
Comment (simonsays1980) [reply rllib/offline/offline_env_runner.py]: I am open to change that. Usually I prefer lower-case folder names.

File: rllib/examples/offline_rl/cartpole_recording.py (hunk 4 above)
Comment (simonsays1980) [reply rllib/examples/offline_rl/cartpole_recording.py:175]: Lowered range to 10. We do not need that much evaluation.

File: rllib/offline/tests/test_offline_rl_stateful.py (hunk 5 above)
Comment (simonsays1980) [reply rllib/offline/tests/test_offline_rl_stateful.py:114]: I use now a while loop to find one - there are multiple in the data set for sure.

File: rllib/offline/offline_env_runner.py (hunk 2 above)
Comment (pseudo-rnd-thoughts) [reply rllib/offline/offline_env_runner.py]: ahh this is for folder names, maybe lower case makes more sense then

File: rllib/offline/offline_prelearner.py (hunk 8 above)
Comment (pseudo-rnd-thoughts) [reply rllib/offline/offline_prelearner.py:305]: I had forgotten about our ORL not support MA yet. Ignore this comment then

File: rllib/offline/offline_prelearner.py (hunk 9 above)
Comment (simonsays1980) [reply rllib/offline/offline_prelearner.py:350]: 1. Should be `0` to ensure we are not using the default in the buffer which is `64`. We need to ensure that this fits the module. The default is also related to DreamerV3. Let's put our heads together in a call to think about what makes most sense here and collect cases.
3. I have not yet thought about all cases - open to discussion - but would think there could be stateful modules that do not rely on sequences but have some other form of memory/state. The `max_seq_len` defines here the sequence length that will be sampled and these non-sequence, but stateful modules could need only a simple 1-step sample.

File: rllib/offline/offline_env_runner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (simonsays1980) [reply rllib/offline/offline_env_runner.py]: The thing is, that we can have local `EnvRunner`s that have no `env` (AlgorithmConfig.create_env_on_local_env_runner=False`). In this case, however, sampling will never be triggered on this EnvRunner. 

I will place an error message inside the `sample` method such that there is no sampling on `EnvRunner`s without `env`.

File: rllib/offline/offline_prelearner.py (hunk 9 above)
Comment (pseudo-rnd-thoughts) [reply rllib/offline/offline_prelearner.py:350]: Ok that makes sense. Should we have a todo note here to resolve in the future if requested?

File: rllib/algorithms/marwil/torch/marwil_torch_learner.py [hunk 11]
```diff
@@ -39,10 +39,16 @@ def compute_loss_for_module(
...
+            mask = batch[Columns.LOSS_MASK].clone()
+            # Check, if a burn-in should be used to recover from a poor state.
+            if self.config.burnin_len > 0:
+                # Train only on the timesteps after the burn-in period.
+                mask[:, : self.config.burnin_len] = False
+            num_valid = torch.sum(mask)
 
             def possibly_masked_mean(data_):
-                return torch.sum(data_[batch[Columns.LOSS_MASK]]) / num_valid
+                return torch.sum(data_[mask]) / num_valid
```
[the remark this answers is not in the document]
Comment (simonsays1980) [reply rllib/algorithms/marwil/torch/marwil_torch_learner.py:51]: Added a validation in `MARWILConfig.training()`

File: rllib/offline/tests/test_offline_env_runner.py [hunk 12]
```diff
@@ -203,6 +203,33 @@ def test_offline_env_runner_compress_columns(self):
...
+    @staticmethod
+    def _get_dir_name(offline_env_runner):
+        if offline_env_runner.env:
+            # Set the subdir (environment specific).
+            if isinstance(offline_env_runner.env, str):
+                # `env` is a string.
+                offline_env_runner.subdir_path = offline_env_runner.env.lower()
+            else:
+                # `env`` is a class or callable we use its class name.
+                offline_env_runner.subdir_path = offline_env_runner.env.unwrapped.envs[
```
Comment (pseudo-rnd-thoughts) [inline rllib/offline/tests/test_offline_env_runner.py]: This will fail for vectorized environments like IsaacLab or PufferDrive. Should be able to test with CartPole-v1 and `gym_vectorized_mode=VectorizeMode.VECTOR_ENTRY_POINT`

File: rllib/offline/offline_env_runner.py [hunk 13]
```diff
@@ -57,8 +57,26 @@ def __init__(self, *, config: AlgorithmConfig, **kwargs):
...
-        self.subdir_path = self.config.env.lower()
+
+        if self.env:
+            # Set the subdir (environment specific).
+            if isinstance(self.env, str):
+                # `env` is a string.
+                self.subdir_path = self.env.lower()
+            else:
+                # `env`` is a class or callable we use its class name.
+                self.subdir_path = self.env.unwrapped.envs[
```
Comment (pseudo-rnd-thoughts) [inline rllib/offline/offline_env_runner.py]: This will only work if the vectorize mode is SYNC, both ASYNC and VECTOR_ENTRY_POINT will fail.

This code will allow you to access the environment name for each of the modes
```python
import gymnasium as gym

if __name__ == "__main__":
    sync_envs = gym.make_vec("CartPole-v1", vectorization_mode=gym.VectorizeMode.SYNC)
    print(f'{sync_envs=}, name={sync_envs.unwrapped.envs[0].unwrapped.__class__.__name__}')

    async_envs = gym.make_vec("CartPole-v1", vectorization_mode=gym.VectorizeMode.ASYNC)
    print(f'{async_envs=}, name={async_envs.unwrapped.get_attr("unwrapped")[0].__class__.__name__}')  # this assumes that the environment is picklable.

    vector_entry_point_envs = gym.make_vec("CartPole-v1", vectorization_mode=gym.VectorizeMode.VECTOR_ENTRY_POINT)
    print(f'{vector_entry_point_envs=}, name={vector_entry_point_envs.unwrapped.__class__.__name__}')
```

**Review by pseudo-rnd-thoughts — APPROVED** [review]
I only have a single comment about collecting the subdir path for different vector environments. If you could make this change (and test locally if all these vectorized mode work), then LGTM
